File: src/source.jl
"""
    AbstractSource

The abstract supertype for custom data sources. Implement a subtype of `AbstractSource` if
`Source{S}` is not sufficient for your source requirements (e.g. your source has additional
information besides the path, such as encoding parameters, compression indicator, etc, that
needs to be associated with each instance).

# Extended help

All subtypes of `AbstractSource` **must**:

- have a `path` field or extend the `sourcepath` function.
- accept a single string, the absolute path of the source file, in the constructor.

Minimum method definitions necessary for complete functionality of a subtype of
`AbstractSource` (i.e. `Source{S}` or otherwise):
- [`readsource`](@ref)
- [`sourcepath`](@ref) (if defining a new subtype of `AbstractSource`)

Additional methods to improve user experience and/or additional functionality
- [`readsegment`](@ref)
- [`generatesource`](@ref) (if enabling `requiresource!` generation)
- [`dependencies`](@ref) (if defining a `generatesource` method)
- [`srcext`](@ref)
- [`srcname_default`](@ref)

## Example

```julia
struct MySource <: AbstractSource
    path::String
end

DatasetManager.readsource(src::MySource) = customreadfunc(sourcepath(src))
DatasetManager.readsegment(seg::Segment{MySource}) = readsource(seg.source)[start:finish]
```
"""
abstract type AbstractSource end

(::Type{S})() where S <: AbstractSource = S(tempname()*srcext(S))

"""
    Source{S}(path)

A basic source, where `S` can be any singleton or existing type
"""
struct Source{S} <: AbstractSource
    path::String
end

Source{S}() where S = Source{S}(tempname()*srcext(Source{S}))

"""
    sourcepath(src) -> String

Return the absolute path to the `src`.
"""
sourcepath(src::AbstractSource) = src.path::String

"""
    srcext(src::Union{S,Type{S}})

    Actual file extension or default file extension for a `src` or `src` type. Period (.) should be the first letter.
"""
srcext(src::AbstractSource) = splitext(sourcepath(src))[2]
srcext(::Type{<:AbstractSource}) = ""

"""
    readsource(src::S; kwargs...) where {S <: AbstractSource}

Read the source data from file.
"""
function readsource end

struct UnknownDeps end

"""
    dependencies(src::Union{S,Type{S}})

Get the sources that `src` depends on to be generated by `generatesource`.
"""
dependencies(::S) where S <: AbstractSource = dependencies(S)
dependencies(::Type{<:AbstractSource}) = UnknownDeps()

"Default name for a source in `Trial.sources`"
srcname_default(::S) where S <: AbstractSource = srcname_default(S)
srcname_default(s) = string(s)

struct MissingSourceError <: Exception
    msg::String
end

function Base.showerror(io::IO, e::MissingSourceError)
    print(io, "MissingSourceError: ")
    print(io, e.msg)
end

"""
    requiresource!(trial, src::Union{S,Type{S}}; force=false, deps, kwargs...) -> nothing
    requiresource!(trial, name::String => src::Union{S,Type{S}}; force=false, deps, kwargs...)

Require a source `src` in `trial`, generate `src` if not present, or throw an error if `src`
is not present and cannot be generated. Returns `nothing`. Keyword argument `deps` defaults to
`dependencies(src)`, but can be manually specified as a `name => src::AbstractSource` pair
per dependency if there exist multiple sources that would match a dependent source type.
`kwargs...` are passed on to the `generatesource` function for `src`.

# Keyword arguments
- `force=false`: Force generating `src`, even if it already exists
- `deps=dependencies(src)`: Manually specify particular dependencies

# Examples
```julia
julia> dependencies(Source{Events})
(Source{C3DFile},)

julia> requiresource!(trial, Source{Events})

julia> requiresource!(trial, Source{Events}; force=true, deps=("mainc3d" => Source{C3DFile}))

```
"""
function requiresource!(trial, src::AbstractSource, parent=nothing; kwargs...)
    requiresource!(trial, srcname_default(src) => src, parent; kwargs...)
end

function requiresource!(trial, src::Type{<:AbstractSource}, parent=nothing; kwargs...)
    requiresource!(trial, srcname_default(src) => src(), parent; kwargs...)
end

function requiresource!(trial, name::Regex)
    requiresource!(trial, name => Source{Nothing}(""), nothing; force=false, deps=UnknownDeps())
end

function requiresource!(trial, namesrc::Pair, parent=nothing;
    force=false, deps=dependencies(namesrc.second), kwargs...
)
    name, src = namesrc
    if !force
        if (parent !== nothing && hassource(trial, typeof(src))) ||
            hassource(trial, name) || hassource(trial, src)
            return nothing
        elseif isfile(sourcepath(src))
            sources(trial)[name] = src
            return nothing
        end
    end

    if deps isa UnknownDeps
        throw(MissingSourceError("unable to generate missing source $src for $trial"))
    else
        foreach(deps) do reqsrc
            requiresource!(trial, reqsrc, src; force=false)
        end
    end
    _src = generatesource(trial, src, deps; kwargs...)
    isfile(sourcepath(_src)) ||
        throw(MissingSourceError("failed to generate source $name => $_src for $trial"))

    sources(trial)[name] = _src

    return nothing
end

"""
    generatesource(trial, src::Union{S,Type{S}}, deps; kwargs...) where S <: AbstractSource -> newsrc::typeof(src)

Generate source `src` using dependent sources `deps` from `trial`. Returns a source of the same
type as `src`, but is not required to be exactly equal to `src` (i.e. a different
`sourcepath(newsrc)` is acceptable).
"""
function generatesource end

generatesource(trial, ::Type{S}, deps; kwargs...) where S <: AbstractSource =
    generatesource(trial, S(), deps; kwargs...)

File: matlab/Source.m
classdef Source
    % SOURCE  A type representing a source of data in a particular file format
    %
    %   src = Source(path)
    %
    % # Input arguments
    %
    % - `path`: The absolute path to the source file
    %
    % # Examples
    %
    % ```matlab
    % src = Source('/path/to/source/file')
    % ```
    %
    % # Creating custom sources
    %
    % All subtypes of `Source` **must**:
    %
    % - have a `path` field
    % - have at least these two constructors:
    %   - empty constructor
    %   - single argument constructor accepting a char vector of an absolute path
    %
    % `Source` subtypes **should**:
    %
    % - have a [`Source.readsource`](@ref) method
    %
    % `Source` subtypes **may** implement these additional methods to improve user
    % experience and/or enable additional funcitonality:
    %
    % - [`Source.readsegment`](@ref)
    % - [`Source.generatesource`](@ref) (if enabling `requiresource!` generation)
    % - [`Source.dependencies`](@ref) (if defining a `generatesource` method)
    % - [`Source.srcext`](@ref)
    % - [`Source.srcname_default`](@ref)

    properties
        path(1,:) char
    end

    methods
        function obj = Source(path)
            if nargin > 0
                obj.path = path;
            else
                obj.path = tempname;
            end
        end

        function bool = eq(obj, y)
            bool = strcmp(class(obj), class(y));
            if bool
                bool = bool & strcmp(obj.path, y.path);
            end
        end

        function ext = srcext(obj)
            % SRCEXT  Get the file extension for a source.
            %
            %   ext = srcext(src)
            %
            % # Input arguments
            %
            % - `src`: Can be the class of a source subtype or an instance of a source. If
            %   `src` is a class, the default extension for that `src < Source` class will be
            %   returned; if `src` is an instance of a source, then the actual extension for
            %   that `src` will be returned, regardless of whether it matches the default
            %   extension for that class or not.
            %
            % # Examples
            %
            % ```matlab
            % % assuming the existence of a class `GaitEvents` which subtypes `Source`
            % src = GaitEvents('/path/to/file.tsv')
            % ext = srcext(GaitEvents) % == '.csv'
            % ext = srcext(src) % == '.tsv'
            % ```
            %
            % # Implementation notes
            %
            % When defining a method for a custom class subtyping `Source`, the period
            % should be included in the extension.
            %
            % ## Example implementation
            %
            % ```matlab
            % function ext = srcext(obj)
            %     %% Calling the default `srcext` method for `Source` will return the actual
            %     %% extension for `< Source` instances
            %     ext = srcext@Source(obj);
            %
            %     if isempty(ext)
            %         ext = '.ext';
            %     end
            % end
            % ```

            if nargin == 0
                ext = '';
            else
                [~,~,ext] = fileparts(obj.path);
            end
        end

        function readsource(obj)
            % READSOURCE  Read the source data from file.
            %
            %   data = readsource(src)
            %

            error('Error: A `readsource` function has not been implemented yet for %s', ...
                class(obj))
        end

        function generatesource(obj, trial, deps, varargin)
            % GENERATESOURCE  Generate a source `src` for `trial` using the dependent
            % sources `deps`. Returns a source of the same class as `src`, but is not
            % required to be exactly equal to `src` (i.e. a different `src.path`).
            %
            %   newsrc = generatesource(src, trial, deps)
            %   newsrc = generatesource(src, trial, deps, Name, Value)
            %
            % # Input arguments
            %
            % - `src`: A source that may or may not exist yet (e.g. if `'Force'` was set to
            %   true in `requiresource`).
            % - `trial`: The trial that `src`/`newsrc` is being generated for/from.
            % - `deps`: The sources that `src` depends on to be generated (i.e. sources
            %   which contain data which is used to generate the data comprising `newsrc`)
            %
            % # Name-Value arguments
            %
            % Any name-value arguments will be specific to a particular Source class
            % subtype. **N.B.**: `generatesource` will normally be called by
            % `requiresource`, which uses name-value arguments `'Name'`, `'Force'`, and
            % `'Dependencies'`. These names are functionally reserved (i.e. not available
            % for use in `generatesource` methods), as they will not be passed on to
            % `generatesource` by `requiresource`.

            error('Error: A `generatesource` function has not been implemented yet for %s', ...
                class(obj))
        end

        function deps = dependencies(obj)
            % DEPENDENCIES  Get the sources that `src` depends on to be generated with
            % `generatesource`. By default sources are assumed not to be automatically
            % generatable, therefore `dependencies` returns false unless a `dependencies`
            % method has been created for a specific class subtyping `Source`.
            %
            %   deps = dependencies(src)

            deps = false;
        end

        function name = srcname_default(obj)
            % SRCNAME_DEFAULT  Get the default name for a source of a given `< Source` class
            %
            %   name = srcname_default(src)

            name = class(obj);
        end

    end

end
